engine/src/main/java/org/terasology/rendering/dag/nodes/ShadowMapNode.java

```diff
@@ -72,86 +77,123 @@
+
+        requiresCondition(() -> renderingConfig.isDynamicShadows());
+        renderingConfig.subscribe(RenderingConfig.DYNAMIC_SHADOWS, this);
 
-        // TODO: fix them
         addDesiredStateChange(new BindFBO(SHADOW_MAP, shadowMapResolutionDependentFBOs));
         addDesiredStateChange(new SetViewportToSizeOf(SHADOW_MAP, shadowMapResolutionDependentFBOs));
+        addDesiredStateChange(new EnableMaterial("engine:prog.shadowMap"));
     }
 
+    private float calculateTexelSize(int shadowMapResolution) {
+        return 1.0f / shadowMapResolution * 2.0f; // the 2.0 multiplier is currently a mystery.
+    }
+
+    /**
+     * Handle changes to the following rendering config properties:
+     *
+     * - DYNAMIC_SHADOWS
+     * - SHADOW_MAP_RESOLUTION
+     *
+     * It assumes the event gets fired only if one of the property has actually changed.
+     *
+     * @param event a PropertyChangeEvent instance, carrying information regarding
+     *              what property changed, its old value and its new value.
+     */
     @Override
-    public void process() {
-        positionShadowMapCamera();
+    public void propertyChange(PropertyChangeEvent event) {
+        if (event.getPropertyName().equals(RenderingConfig.DYNAMIC_SHADOWS)) {
+            super.propertyChange(event);
+        } else if (event.getPropertyName().equals(RenderingConfig.SHADOW_MAP_RESOLUTION)) {
+            int shadowMapResolution = (int) event.getNewValue();
+            texelSize = calculateTexelSize(shadowMapResolution);
+        }
+    }
 
-        // TODO: find an elegant way to fetch isFirstRenderingStageForCurrentFrame
-        // TODO: check these conditions before adding this node inside the DAG, removing this condition completely from process()
-        if (renderingConfig.isDynamicShadows() && worldRenderer.isFirstRenderingStageForCurrentFrame()) {
+    /**
+     * Re-positions the shadow map camera to loosely match the position of the main light (sun, moon), then
+     * writes depth information from that camera into a depth buffer, to be used later to create shadows.
+     *
+     * The loose match is to avoid flickering: the shadowmap only moves in steps while the main light actually
+     * moves continuously.
+     *
+     * This method is executed within a NodeTask in the Render Tasklist, but its calculations are executed
+     * only once per frame. I.e. in VR mode they are executed only when the left eye is processed. This is
+     * done in the assumption that we do not need to generate and use a shadow map for each eye as it wouldn't
+     * be noticeable.
+     */
+    @Override
+    public void process() {
+        // TODO: remove this IF statement when VR is handled via parallel nodes, one per eye.
+        if (worldRenderer.isFirstRenderingStageForCurrentFrame()) {
             PerformanceMonitor.startActivity("rendering/shadowMap");
+            positionShadowMapCamera();
```

Comment:
@indianajohn: this move makes sure that the shadow camera is positioned only once per frame instead of once per eye in VR mode. Does this look ok to you? Or do we need to render a different shadow map for each eye? (which would require some modification to the ShadowMapNode I imagine).